packages/runtime/src/state/react-page.ts

```diff
@@ -168,6 +169,116 @@ export function getElementId(state: State, documentKey: string, elementKey: stri
+  fileIds: Map<string, string>
+  typographyIds: Map<string, string>
+  tableIds: Map<string, string>
+  tableColumnIds: Map<string, string>
+  pageIds: Map<string, string>
+  globalElementIds: Map<string, string>
+  globalElementData: Map<string, Documents.ElementData>
+}
+
+export function createReplacementContext(jsonReplacementContext: {
+  elementHtmlIds: string[]
+  elementKeys: { [s: string]: string }
+  swatchIds: { [s: string]: string }
+  fileIds: { [s: string]: string }
+  typographyIds: { [s: string]: string }
+  tableIds: { [s: string]: string }
+  tableColumnIds: { [s: string]: string }
+  pageIds: { [s: string]: string }
+  globalElementIds: { [s: string]: string }
+  globalElementData: { [s: string]: Documents.ElementData }
+}): ReplacementContext {
+  const rc: ReplacementContext = {
+    elementHtmlIds: new Set(jsonReplacementContext.elementHtmlIds),
+    elementKeys: new Map(
+      jsonReplacementContext.elementKeys && Object.entries(jsonReplacementContext.elementKeys),
+    ),
+    swatchIds: new Map(
+      jsonReplacementContext.swatchIds && Object.entries(jsonReplacementContext.swatchIds),
+    ),
+    fileIds: new Map(
+      jsonReplacementContext.fileIds && Object.entries(jsonReplacementContext.fileIds),
+    ),
+    typographyIds: new Map(
+      jsonReplacementContext.typographyIds && Object.entries(jsonReplacementContext.typographyIds),
+    ),
+    tableIds: new Map(
+      jsonReplacementContext.tableIds && Object.entries(jsonReplacementContext.tableIds),
+    ),
+    tableColumnIds: new Map(
+      jsonReplacementContext.tableColumnIds &&
+        Object.entries(jsonReplacementContext.tableColumnIds),
+    ),
+    pageIds: new Map(
+      jsonReplacementContext.pageIds && Object.entries(jsonReplacementContext.pageIds),
+    ),
+    globalElementIds: new Map(
+      jsonReplacementContext.globalElementIds &&
+        Object.entries(jsonReplacementContext.globalElementIds),
+    ),
+    globalElementData: new Map(
+      jsonReplacementContext.globalElementData &&
+        Object.entries(jsonReplacementContext.globalElementData),
+    ),
+  }
+
+  return rc
+}
+
+export type CopyContext = {
+  replacementContext: ReplacementContext
+  copyElement: (node: Documents.Element) => Documents.Element
+}
+
+export function copyElementTree(
+  state: State,
+  elementTree: Documents.ElementData,
+  replacementContext: ReplacementContext,
+) {
+  /*
+   * This is structured a bit weird.
+   *
+   * This is done so that we can pass a callable function into some of the copy functions
+   * themselves, to enable mutual recursion.
+   *
+   * Consider the slot control. It has to iterate through its elements, and for each of them,
+   * call some version of the below function.
+   *
+   * That is how the recursing through the tree happens.
+   */
+  function copyElementTreeNode(state: State, replacementContext: ReplacementContext) {
+    return function (node: Documents.Element) {
+      if ('props' in node) {
```

Comment:
Seems like we also want to handle element references and not just element data, right? For example, if it's an element reference we want to replace the global element ID with the one in the replacement context?

Also, we've got guard functions to refine element references and element data. I believe it's `isElementReference` and `isElementData`. They're set up so that if you do `if(isElementReference(anyValue))` the `if` block refines `anyValue` to the checked type.

Reply:
Yeah I'm aware of that, gonna handle this when I get to global elements